packages/transport-webrtc/src/util.ts

```diff
@@ -1,8 +1,59 @@
+import { logger } from '@libp2p/logger'
 import { detect } from 'detect-browser'
+import pDefer from 'p-defer'
+
+const log = logger('libp2p:webrtc:utils')
 
 const browser = detect()
 export const isFirefox = ((browser != null) && browser.name === 'firefox')
 
 export const nopSource = async function * nop (): AsyncGenerator<Uint8Array, any, unknown> {}
 
 export const nopSink = async (_: any): Promise<void> => {}
+
+export function drainAndClose (channel: RTCDataChannel, channelCloseDelay: number): void {
+  if (channel.readyState !== 'open') {
+    return
+  }
+
+  void Promise.resolve()
+    .then(async () => {
+      // wait for bufferedAmount to become zero
+      if (channel.bufferedAmount > 0) {
+        const deferred = pDefer()
+
+        channel.bufferedAmountLowThreshold = 0
+        channel.addEventListener('bufferedamountlow', () => {
+          deferred.resolve()
+        })
+
+        await deferred.promise
+      }
+    })
+    .then(async () => {
+      const deferred = pDefer()
+
+      // event if bufferedAmount is zero there can still be unsent bytes
+      const timeout = setTimeout(() => {
+        try {
+          // only close if the channel is still open
+          if (channel.readyState === 'open') {
+            channel.close()
+          }
+
+          deferred.resolve()
+        } catch (err: any) {
+          deferred.reject(err)
+        }
+      }, channelCloseDelay)
+
+      if (timeout.unref != null) {
+        timeout.unref()
+      }
+
+      await deferred.promise
+    })
+    .catch(err => {
+      log.error('error closing outbound stream', err)
```

Comment:
Do you think we should bubble this error up? i.e. throw an `ERR_DATA_CHANNEL` ?

Reply:
I thought about it, but it would make the stream `onEnd` callback need to be async which seems wrong.

From testing the only time this throws is when the datachannel is already closed but this PR guards on the `readyState` before closing so the `catch` is really just here to catch edge cases.